crates/futures/src/atomics.rs

```diff
@@ -0,0 +1,351 @@
+// idea here is thus to do as much synchronous work as we can and then otherwise
+// translate notifications of a future's task into "let's poll the future!"
+//
+// This isn't necessarily the greatest future executor in the world, but it
+// should get the job done for now hopefully.
+fn _future_to_promise(future: Box<dyn Future<Item = JsValue, Error = JsValue>>) -> Promise {
+    let mut future = Some(executor::spawn(future));
+    return Promise::new(&mut |resolve, reject| {
+        Package::poll(&Arc::new(Package {
+            spawn: RefCell::new(future.take().unwrap()),
+            resolve,
+            reject,
+            notified: Cell::new(State::Notified),
+            waker: Arc::new(Waker::default()),
+        }));
+    });
+
+    struct Package {
+        // Our "spawned future". This'll have everything we need to poll the
+        // future and continue to move it forward.
+        spawn: RefCell<Spawn<Box<dyn Future<Item = JsValue, Error = JsValue>>>>,
+
+        // The current state of this future, expressed in an enum below. This
+        // indicates whether we're currently polling the future, received a
+        // notification and need to keep polling, or if we're waiting for a
+        // notification to come in (and no one is polling).
+        notified: Cell<State>,
+
+        // Our two callbacks connected to the `Promise` that we returned to
+        // JavaScript.  We'll be invoking one of these at the end.
+        resolve: Function,
+        reject: Function,
+
+        // Struct to wake a future
+        waker: Arc<Waker>,
+    }
+
+    // The possible states our `Package` (future) can be in, tracked internally
+    // and used to guide what happens when polling a future.
+    enum State {
+        // This future is currently and actively being polled. Attempting to
+        // access the future will result in a runtime panic and is considered a
+        // bug.
+        Polling,
+
+        // This future has been notified, while it was being polled. This marker
+        // is used in the `Notify` implementation below, and indicates that a
+        // notification was received that the future is ready to make progress.
+        // If seen, however, it probably means that the future is also currently
+        // being polled.
+        Notified,
+
+        // The future is blocked, waiting for something to happen. Stored here
+        // is a self-reference to the future itself so we can pull it out in
+        // `Notify` and continue polling.
+        //
+        // Note that the self-reference here is an Arc-cycle that will leak
+        // memory unless the future completes, but currently that should be ok
+        // as we'll have to stick around anyway while the future is executing!
+        //
+        // This state is removed as soon as a notification comes in, so the leak
+        // should only be "temporary"
+        Waiting(Arc<Package>),
+    }
+
+    struct Waker {
```

Comment:
@alexcrichton so we can use only one field here, `notified`, and make it `AtomicI32`?